libbpf-rs/src/perf_buffer.rs

```diff
@@ -71,9 +71,13 @@ impl<'a, 'b> PerfBufferBuilder<'a, 'b> {
     }
 
     /// The number of pages to size the ring buffer.
-    pub fn pages(&mut self, pages: usize) -> &mut Self {
-        self.pages = pages;
-        self
+    pub fn pages(self, pages: usize) -> PerfBufferBuilder<'a, 'b> {
+        PerfBufferBuilder {
+            map: self.map,
+            pages,
+            sample_cb: self.sample_cb,
+            lost_cb: self.lost_cb,
+        }
```

Comment:
I don't think this is quite right - to clarify, what is the intention with your proposed API changes to pages? The original version of pages looks fine to me, someone can set the pages of the perf buffer through calling something like:

```
let perf = 
   PerfBufferBuilder::new(skel.maps_mut().events()).pages(5)
   .sample_cb(handle_event).lost_cb(handle_lost_events).build()?;
```

Reply:
Without my fix your line will produces this error for me:
   Compiling libbpf-rs v0.15.0 (/.../libbpf-rs/libbpf-rs)
   Compiling rust_bpf_agent v0.1.0 (/.../rust_bpf_agent)
error[E0507]: cannot move out of a mutable reference
   --> src/main.rs:116:16
    |
116 |       let perf = PerfBufferBuilder::new(skel.maps_mut().events())
    |  ________________^
117 | |         .sample_cb(handle_event)
118 | |         .lost_cb(handle_lost_events)
119 | |         .pages(1024)
120 | |         .build()?;
    | |________________^ move occurs because value has type `PerfBufferBuilder<'_, '_>`, which does not implement the `Copy` trait

For more information about this error, try `rustc --explain E0507`.
error: could not compile `rust_bpf_agent` due to previous error

Reply:
While we are talking about this, is there a way I can pass PerfBuffer loop to a different thread? (Currently it does not implement Send)
Think of a case where the eBPF programs sends A LOT of messages and you want to keep fetching events and put it in some channel to do further more cpu intensive processing, but you dont want the ebpf map to get full...

Reply:
Looking at lost_cb definition it differs from what is done in pages, so the change looks in line with other setters. I'm surprised this was never caught before, don't we have some example or test that uses PerfBuffer? Can you please add one?

Reply:
> While we are talking about this, is there a way I can pass PerfBuffer loop to a different thread? (Currently it does not implement Send) Think of a case where the eBPF programs sends A LOT of messages and you want to keep fetching events and put it in some channel to do further more cpu intensive processing, but you dont want the ebpf map to get full...

Not clear what exactly you are sending in this case, but keep in mind that as it is designed in libbpf right now, PerfBuffer is single-consumer and poll has to be called from single thread. Also keep in mind that each per-CPU buffer has to be single-consumer. The only wiggle room is to do poll on different per-cpu buffers on different threads.

But please start a separate issue to discuss this use case.

Reply:
I dont think it's such a big deal, one could always wrap *Skel in it's own struct and implement Send for it
But in general I dont understand why the *Skel cannot be Send, unless I'm missing something

Reply:
Probably no one ever needed it. I'm not too familiar with Sync and Send, but if there are no issues with the way PerfBuffer is implemented, then yeah, we should implement those traits.